crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -108,263 +91,130 @@ impl CheckpointExecutor {
-                                last_checkpoint=checkpoint.sequence_number(),
-                                "Reached end of epoch",
-                            );
-                            self.checkpoint_store
-                                .update_highest_executed_checkpoint(&checkpoint)
-                                .unwrap();
-                            self.metrics.last_executed_checkpoint.set(checkpoint.sequence_number() as i64);
-
-                            // be extra careful to ensure we don't have orphans
-                            assert!(
-                                pending.is_empty(),
-                                "Pending checkpoint execution buffer should be empty after processing last checkpoint of epoch",
-                            );
-
-                            return self.create_committee_object(checkpoint, committee);
-                        }
-                    }
+                Some(Ok(checkpoint)) = pending.next() => {
+                    self.finished_executing_checkpoint(&checkpoint);
+                    highest_executed = Some(checkpoint);
                 }
                 // Check for newly synced checkpoints from StateSync.
-                received = self.mailbox.recv() => match received {
-                    Ok(checkpoint) => {
-                        debug!(
-                            "Received new synced checkpoint message for checkpoint {:?}",
-                            checkpoint.sequence_number(),
-                        );
-                        self.latest_synced_checkpoint = Some(checkpoint);
-                    }
-                    // In this case, messages in the mailbox have been overwritten
-                    // as a result of lagging too far behind. In this case, we need to
-                    // nullify self.latest_synced_checkpoint as the latest synced needs to
-                    // be read from the watermark
-                    Err(RecvError::Lagged(num_skipped)) => {
-                        self.latest_synced_checkpoint = None;
-
-                        warn!(
-                            "Checkpoint Execution Recv channel overflowed {:?} messages",
-                            num_skipped,
-                        );
-                        self.metrics
-                            .checkpoint_exec_recv_channel_overflow
-                            .inc_by(num_skipped);
-                    }
-                    Err(RecvError::Closed) => {
-                        panic!("Checkpoint Execution Sender (StateSync) closed channel unexpectedly");
-                    }
-                },
+                _ = self.mailbox.recv() => (),
             }
         }
     }
 
+    /// Post processing and plumbing after we executed a checkpoint. This function is guaranteed
+    /// to be called in the order of checkpoint sequence number.
+    fn finished_executing_checkpoint(&self, checkpoint: &VerifiedCheckpoint) {
```

Comment:
nit: can we call this something like `process_executed_checkpoint`? The current title feels like it should return a ` bool`